File: M6/testcontainers-python-postgresql/tests/3_generic_container_test.py
import os
import pytest
import psycopg2
import psycopg2.extras
from testcontainers.core.container import DockerContainer
from testcontainers.core.waiting_utils import wait_for_logs
from .db_client import DBClient

MOCK_DATA = [
    {'id': 1, 'name': 'Alice', 'age': 30},
    {'id': 2, 'name': 'Bob', 'age': 24},
    {'id': 3, 'name': 'Charlie', 'age': 28},
    {'id': 4, 'name': 'Diana', 'age': 35},
    {'id': 5, 'name': 'Eve', 'age': 22},
    {'id': 6, 'name': 'Frank', 'age': 33},
    {'id': 7, 'name': 'Grace', 'age': 27},
    {'id': 8, 'name': 'Hank', 'age': 31},
    {'id': 9, 'name': 'Ivy', 'age': 29},
    {'id': 10, 'name': 'Jack', 'age': 26},
]

@pytest.fixture(scope="module")
def db_connection():
    postgres_user = "test"
    postgres_password = "test"
    postgres_db = "test"
    postgres_port = 5432

    with DockerContainer("postgres:15-alpine") \
            .with_env("POSTGRES_USER", postgres_user) \
            .with_env("POSTGRES_PASSWORD", postgres_password) \
            .with_env("POSTGRES_DB", postgres_db) \
            .with_exposed_ports(postgres_port) \
            .with_volume_mapping(os.path.join(os.path.dirname(__file__), 'data.sql'), '/docker-entrypoint-initdb.d/data.sql') as postgres_container:
        
        postgres_container.start()
        wait_for_logs(postgres_container, "database system is ready to accept connections", timeout=60)

        host = postgres_container.get_container_host_ip()
        port = postgres_container.get_exposed_port(postgres_port)

        dsn = f"postgresql://{postgres_user}:{postgres_password}@{host}:{port}/{postgres_db}"
        db_client = DBClient(dsn)
        db_client.connect()
        yield db_client
        db_client.close()

def test_should_fetch_all_mock_persons_from_database(db_connection):
    result_dicts = db_connection.fetch_all("SELECT * FROM persons ORDER BY id")
    assert len(result_dicts) == len(MOCK_DATA)
    assert result_dicts == MOCK_DATA

def test_should_fetch_persons_older_than_30(db_connection):
    result_dicts = db_connection.fetch_all("SELECT * FROM persons WHERE age > %s ORDER BY id", [30])
    expected = [p for p in MOCK_DATA if p['age'] > 30]
    assert result_dicts == expected

File: M6/testcontainers-vitest-postgresql/test/3.generic-container.test.ts
import { expect, test, beforeAll, afterAll } from 'vitest';
import pg, { Client } from 'pg';
import { GenericContainer, StartedTestContainer, Wait } from 'testcontainers';
import path from 'path';

let client: typeof Client;
let container: StartedTestContainer;

beforeAll(async () => {
	const sqlFileHostPath = path.resolve(__dirname, 'data.sql');
	const sqlFileContainerPath = '/docker-entrypoint-initdb.d/01-init.sql';

	// Use GenericContainer to start Postgres image, expose port 5432, wait for log line, copy init SQL
		container = await new GenericContainer('postgres:15-alpine')
		.withEnvironment({
			POSTGRES_PASSWORD: 'testpwd',
			POSTGRES_USER: 'testuser',
			POSTGRES_DB: 'testdb'
		})
		.withExposedPorts(5432)
		.withCopyFilesToContainer([
			{ source: sqlFileHostPath, target: sqlFileContainerPath }
		])
			.withWaitStrategy(Wait.forLogMessage('database system is ready to accept connections'))
		.start();

	const host = container.getHost();
	const port = container.getMappedPort(5432);
		// Retry loop to handle race where init scripts not yet fully applied
		for (let attempt = 1; attempt <= 5; attempt++) {
			try {
				client = new Client({
					host,
					port,
					user: 'testuser',
					password: 'testpwd',
					database: 'testdb'
				});
				await client.connect();
				// simple probe
				await client.query('SELECT 1');
				break;
			} catch (e) {
				if (attempt === 5) throw e;
				await new Promise(r => setTimeout(r, 500 * attempt));
			}
		}
});

afterAll(async () => {
	await client.end();
	await container.stop();
});

test('generic container: all persons', async () => {
	const res = await client.query('SELECT * FROM persons ORDER BY id');
	expect(res.rowCount).toBe(10);
	expect(res.rows.map(r => r.name)).toEqual(['Alice','Bob','Charlie','Diana','Eve','Frank','Grace','Hank','Ivy','Jack']);
});

test('generic container: age > 30', async () => {
	const res = await client.query('SELECT * FROM persons WHERE age > $1 ORDER BY id', [30]);
	expect(res.rows.map(r => r.id)).toEqual([4,6,8]);
});
